keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,837 @@
+for windows, or if we support TCP/IP in the future.
+
+### Caching
+
+The connection from the client to the proxy can be cached in the client based
+on the cluster details, the existing HTTP_PROXY configuration, the exec-proxy
+URL, and the certificates used for mTLS.
+
+### Proxy Shutdown
+
+In the case where the client dies unexpectedly, due to a crash or SIGKILL, the
+proxy should also die.  To handle this, the proxy will attempt to take a file
+lock that the parent holds and quit if it ever successfully takes the lock.
+
+Alterntatively, the proxy could quit after a certain amount of idle time.  For
+example, proxies that are idle for more than 5 minutes will be required to
+quit.  Idle quit is simpler, but the lock approach handles the kubectl use case
+better, so it is preferred.
+
+### API
+
+The communication over stdin and stdout with the exec'd binary will be done
+with using the existing ExecCredentialSpec and ExecCredentialStatus objects.
+
+```go
+// ExecCredentialSpec holds request and runtime specific information provided by
+// the transport.
+type ExecCredentialSpec struct {
+	// Cluster contains information to allow an exec plugin to communicate with the
+	// kubernetes cluster being authenticated to. Note that Cluster is non-nil only
+	// when provideClusterInfo is set to true in the exec provider config (i.e.,
+	// ExecConfig.ProvideClusterInfo).
+	// +optional
+	Cluster *Cluster `json:"cluster,omitempty"`
+
+	// Interactive declares whether stdin has been passed to this exec plugin.
+	Interactive bool `json:"interactive"`
+}
+
+// ExecCredentialStatus holds credentials for the transport to use.
+//
+// Token and ClientKeyData are sensitive fields. This data should only be
+// transmitted in-memory between client and exec plugin process. The Exec plugin
+// itself should at least be protected via file permissions.
+type ExecCredentialStatus struct {
+	// ExpirationTimestamp indicates a time when the provided credentials expire.
+	// +optional
+	ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
+	// Token is a bearer token used by the client for request authentication.
+	Token string `json:"token,omitempty" datapolicy:"token"`
+	// PEM-encoded client TLS certificates (including intermediates, if any).
+	ClientCertificateData string `json:"clientCertificateData,omitempty"`
+	// PEM-encoded private key for the above certificate.
+	ClientKeyData string `json:"clientKeyData,omitempty" datapolicy:"security-key"`
+
+	// New:
+
+	// ProxyConfig is the config returned by the exec-proxy
+	// +optional
+	ProxyConfig *ProxyConfig `json:"proxyConfig,omitempty"`
+}
+
+
+// ProxyConfig configures the client to proxy requests via the returned address.
+type ProxyConfig struct {
+	// URL is the location of the exec proxy listener.  Supports unix domain
+	// socket locations in the format unix:///path/to/socket.
+	URL string `json: "url"`
+	// CertificateAuthorityData is required if an exec-proxy is used.
```

Comment:
Is this always necessary? Why?

Reply:
Because the connection between the proxy and client-go is always over mutual TLS.  We do not rely on the UDS to provide any security guarantees especially since they behave differently on windows.

Reply:
there are scenarios that mtls is not necessary. should we leave the decision to the users?

Reply:
If we leave the decision to the users, then we need to allow them to use the socket permissions and the directory permissions in which the socket resides to provide security guarantees.

Reply:
> socket permissions

We use socket permissions all over the place (docker socket, KMS plugin, device plugin, CSI driver)... without issue. It's a well understood permission model. While not particularly critical for kubectl, if efficiency ever becomes a concern (e.g. usage expands to cluster components like kubelet) then we will want to revisit the mTLS requirement.

Reply:
Using socket permissions on server components that can be tightly controlled is not the same as using it on end user desktop that may have more than one user and questionable security controls (and an unbounded set of custom software).  We can certainly measure the overhead of mTLS but I suspect it is negligible on any modern hardware.  Unlike the API server, this proxy can do connection level mTLS checks instead of per request mTLS checks, which will further reduce any issues.  And again, the reference implementation will do this all for you - you only have to specify the business logic that is custom to your proxy, not this boilerplate.

Reply:
Let's remove this comment for now and work it out over the implementation.